### PR TITLE
Upgrade to Linux 6.6.18

### LINKED ISSUE #181: Rename ttyS* devices to correspond to node numbers
Currently, the ttyS* devices don't correspond to the node they're connected to, e.g. `/dev/ttyS4` corresponds to node 3. This is not necessary: this is a purpose-built system, and the fact that the BMC's SOC calls a certain UART something doesn't mean that the BMC firmware has to do the same. Renaming them would make things easier on users, who won't have to remember which device corresponds to which node.

I'd like to propose the following mapping:
* `/dev/ttyS0`: for the BMC itself (the connection that's exposed on the Turing Pi as `BMC_UART3`)
* `/dev/ttyS1`: Node 1
* `/dev/ttyS2`: Node 2
* `/dev/ttyS3`: Node 3
* `/dev/ttyS4`: Node 4

### PR DESCRIPTION
**Pull request by CFSworks**
- Unsure about level of compatibility with current `bmcd`
- Completely untested on TPi2.4. All functionality needs to be fully verified before accepting this PR.
  - [x] OTA upgrade works
  - [x] microSD card installation works
  - [x] OTA downgrade to 2.0.x after microSD card installation works
  - [x] CM4 flashing works
  - [x] RK1 flashing works
  - [x] BMC USB port works in device mode
  - [x] BMC USB port works in host mode
  - [x] BMC UART works
  - [x] Keys, LEDs work
  - [x] Serial access to each node works
  - [ ] RTC still works, correct time is set on boot even if network not up
  - [x] For those that have soldered the EMC2305 fan controller, that still works
- Gives up thermal sensor (for now)
- Closes #181

Please cherry-pick/rebase instead of merge.

### CODE REVIEW

**Comment by donaldguy** [comment]
(As someone who just set up their TP2.4 today after getting 32GB RK1s, and is already fighting with network arch problems,) I'd like to put in a word in favor of hoping this big jump opens up #46 (since its past the additions in 5.16 you mentioned [here](https://github.com/turing-machines/BMC-Firmware/issues/46#issuecomment-1556085994) 

I'm not currently seeing any defconfig or device tree to get the hooks into the RTL8370MB ; but ... maybe I'm missing something obvious

I could see a case for relegating the switch driver to an unloaded-by-default-module until e.g. hooks in bmcd are there to make it useful in userspace

but ditto can imagine the presence of the driver might be worth having in whatever round of manual QA is happening here

File: tp2bmc/board/tp2bmc/sun8i-t113s-turing-pi2.dtsi [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+	allwinner,rx-delay-ps = <0>;
+};
+
+&mdio {
+	rtl8201f: ethernet-phy@0 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <0>;
+		reset-gpios = <&pio 4 10 GPIO_ACTIVE_LOW>; /* PE10 */
+	};
+};
```
Comment (donaldguy) [inline tp2bmc/board/tp2bmc/sun8i-t113s-turing-pi2.dtsi:242]: **I don't really speak kernel or hardware or understand what pins are wired to which controller** 

But in a probably vein attempt to be helpful with what I'm requesting I surmise (from:
-  https://www.kernel.org/doc/Documentation/devicetree/bindings/net/dsa/realtek-smi.txt , 
- https://www.kernel.org/doc/Documentation/devicetree/bindings/net/dsa/dsa-port.yaml , 
-  https://github.com/devicetree-org/devicetree-specification/blob/2697c66a59b9177f8ff7813c355cc7aa05f6e40a/source/chapter4-device-bindings.rst#L344  
[and not really helped by 
`https://cdn.discordapp.com/attachments/1107160366888267866/1109336549222514768/2208101101_Realtek-Semicon-RTL8370N-VB-CG_C2803243.pdf?ex=6610cf2b&is=65fe5a2b&hm=11f81f2a2c1a05af62fd449af17536ae4da02a5561e55c91b92cc20f7352df7f&` nor `https://cdn.discordapp.com/attachments/1107160366888267866/1109336548916350996/Realtek_Unmanaged_Switch_ProgrammingGuide.pdf?ex=6610cf2b&is=65fe5a2b&hm=974a489d57215bd6db3ac82a94ed9705397a8b239d7bf2998d246e0ce42bcefe`& ) the broad shape of things **_might_** be something like 

```
&switch {
	compatible = "realtek,rtl8370";
	mdc-gpios = <&pio ? ?? GPIO_ACTIVE_??>;
	mdio-gpios = <&pio 4 ?? GPIO_ACTIVE_???>;
	reset-gpios = <&pio ? ?? GPIO_ACTIVE_??>;

	switch_intc: interrupt-controller {
		interrupt-parent = <&pio ?>;
		interrupts = <1 IRQ_TYPE_LEVEL_LOW>;
		interrupt-controller;
		#address-cells = <0>;
		#interrupt-cells = <1>;
	};

	ports {
		#address-cells = <1>;
		#size-cells = <0>;
		reg = <0>;

		port@0 {
			reg = <0>;
			label = "node1-ethernet0";
			ethernet = <&nodes "node1-en">;
			phy-handle = <&emac>
		};
		port@1 {
			reg = <1>;
			label = "node1-ethernet1";
			ethernet = <&nodes "node1-en">;
			phy-handle = <&emac1>
		};
		port@2 {
			reg = <2>;
			label = "node2-ethernet0";
			ethernet = <&nodes "node2-en">;
			phy-handle = <&emac>
		};
		port@3 {
			reg = <3>;
			label = "node2-ethernet1";
			ethernet = <&nodes "node2-en">;
			phy-handle = <&emac1>
		};
		port@4 {
			reg = <4>;
			label = "node3-ethernet0";
			ethernet = <&nodes "node3-en">;
			phy-handle = <&emac>
		};
                port@5 {
			reg = <5>;
			label = "node3-ethernet1";
			ethernet = <&nodes "node3-en">;
			phy-handle = <&emac1>
		};
                port@6 {
			reg = <6>;
			label = "node4-ethernet0";
			ethernet = <&nodes "node4-en">;
			phy-handle = <&emac>
		};
                port@7 {
			reg = <7>;
			label = "node4-ethernet1";
			ethernet = <&nodes "node4-en">;
			phy-handle = <&emac1>
		};
	};

	mdio {
		compatible = "realtek,smi-mdio";
		#address-cells = <1>;
		#size-cells = <0>;

		emac: phy@0 {
			reg = <0>;
			interrupt-parent = <&switch_intc>;
			interrupts = <0>;
		};
		emac1: phy@1 {
			reg = <1>;
			interrupt-parent = <&switch_intc>;
			interrupts = <1>;
		};
	};
};
```

(but of course the bmc itself is clearly inserted into the loop somewhere... so ... seems doubtful [maybe the 8 ports are 4 nodes + bmc + 2 physical ports + ? ; I kinda thought 8 + 2 would be to ten but those pdfs whomever posted on discord suggest only numbered up to 7 ]

Comment (donaldguy) [reply tp2bmc/board/tp2bmc/sun8i-t113s-turing-pi2.dtsi:242]: though honestly I'm already lost at the one line I thought I got, cause: 
 https://www.kernel.org/doc/Documentation/devicetree/bindings/net/dsa/realtek.yaml (which is more authoritative?) still says to use "realtek,rtl8365mb"

even though https://www.kernel.org/doc/Documentation/devicetree/bindings/net/dsa/realtek-smi.txt lists `"realtek,rtl8370`

[ apparently cause https://patchwork.kernel.org/project/netdevbpf/patch/20220418233558.13541-1-luizluca@gmail.com/ ? ]

Comment (donaldguy) [reply tp2bmc/board/tp2bmc/sun8i-t113s-turing-pi2.dtsi:242]: there is also confusing re-factor of "dsa" specific things vs more generic ethernet switching device tree stuff ; ... I give up [ but probably one of the the more complicated things I guessed at in the middle and then deleted was closer ; but still wrong]

**Comment by dgdonovan** [comment]
Initial test results from my TPiv2.4:
- microSD card installation works.
- OTA downgrade to 2.0.x via via BMC UI with .tpu file after microSD card installation works.
- OTA upgrade to this build via BMC UI with .tpu file works.
- BMC UART works ("CONFIRM" entered through this port to complete BMC firmware programming)
- POWER+GND header wired to chassis switch works (pressed three times in quick succession to "CONFIRM" SD card installation; pressed once to turn all nodes on; pressed once to turn all nodes off)
- RESET+GND header wired to chassis switch works (resets BMC; nodes that were powered on prior to BMC reset are not powered on after BMC reset).
- KEY1 button works (pressed three times in quick succession to "CONFIRM" SD card installation; pressed once to turn all nodes on; pressed once to turn all nodes off)
- BMC_RESET button works (resets BMC; nodes that were powered on prior to BMC reset are not powered on after BMC reset).
- Ethernet LEDs all flash slowly when SD card firmware installation prompts for "CONFIRM". Unlike earlier v2.0.5 firmware versions, the P-LED header is not flashed along with the Ethernet LEDs.
- Ethernet LEDs sequentially flash quickly while firmware installation from SD card occurs. Unlike earlier v2.0.5 firmware versions, the P-LED header is not flashed quickly during firmware installation.
- Ethernet LEDs are repeat a sequence of all flashing twice quickly then pausing briefly when SD card firmware installation completes. Unlike earlier v2.0.5 firmware versions, the P-LED header is not flashed with the Ethernet LEDs.
- When nodes are powered on via either the KEY1 button, the POWER+GND chassis header or any node via the UI, appropriate TPiv2 red node LEDs are on and the S-LED header is on, main PSU power is on and the FAN header is on.
- When nodes are powered off via either the KEY1 button, the POWER+GND chassis header or all nodes via the UI, TPiv2 red node LEDs are off, the S-LED header is off, main PSU power is off and the FAN header is off.
- The P-LED header is turned on for several seconds when the BMC is reset from the UI.
- /var/lib/bmcd/bmcd.bin is not preserved across BMC reboots.
- If the PWM fan header and EMC2301 chip are present, /sys/class/thermal/cooling_device0 and associated fifos are present and function the same as in prior versions of v2.0.5 firmware.

**Comment by svenrademakers** [comment]
Successfully tested using `tp2-firmware-sdcard-v2.0.5-35-g0ea3fd8a.img` :
- flashing RK1
- flashing CM4
- serial access to each node
- BMC-USB-OTG ACM interface

**Comment by svenrademakers** [comment]
created tracking issue for persistency problem: https://github.com/turing-machines/BMC-Firmware/issues/191

**Pull request merged**